src/wasmtime/wasmtime.cc

```diff
@@ -0,0 +1,674 @@
+    } break;
+    case WASM_EXTERN_GLOBAL: {
+      // TODO(mathetake): add support when/if needed.
+      fail(FailState::UnableToInitializeCode,
+           "Failed to load Wasm module due to a missing import: " + std::string(module_name) + "." +
+               std::string(name));
+    } break;
+    case WASM_EXTERN_MEMORY: {
+      assert(memory_ == nullptr);
+      const wasm_memorytype_t *memory_type =
+          wasm_externtype_as_memorytype_const(extern_type); // owned by `extern_type`
+      memory_ = wasm_memory_new(store_.get(), memory_type);
+      imports.push_back(wasm_memory_as_extern(memory_.get()));
+    } break;
+    case WASM_EXTERN_TABLE: {
+      assert(table_ == nullptr);
+      const wasm_tabletype_t *table_type =
+          wasm_externtype_as_tabletype_const(extern_type); // owned by `extern_type`
+      table_ = WasmTablePtr{wasm_table_new(store_.get(), table_type, nullptr)};
+      imports.push_back(wasm_table_as_extern(table_.get()));
+    } break;
+    }
+  }
+
+  if (import_types.size != imports.size()) {
+    return false;
+  }
+
+  instance_ = wasm_instance_new(store_.get(), module_.get(), imports.data(), nullptr);
+  assert(instance_ != nullptr);
+
+  wasm_exporttype_vec_t export_types;
+  wasm_module_exports(module_.get(), &export_types);
+
+  wasm_extern_vec_t exports;
+  wasm_instance_exports(instance_.get(), &exports);
+  assert(exports.size == export_types.size);
+
+  for (size_t i = 0; i < export_types.size; i++) {
+    const wasm_externtype_t *exp_extern_type = wasm_exporttype_type(export_types.data[i]);
+    wasm_extern_t *actual_extern = exports.data[i];
+
+    wasm_externkind_t kind = wasm_extern_kind(actual_extern);
+    assert(kind == wasm_externtype_kind(exp_extern_type));
+    switch (kind) {
+    case WASM_EXTERN_FUNC: {
+      wasm_func_t *func = wasm_func_copy(wasm_extern_as_func(actual_extern));
+      const wasm_name_t *name_ptr = wasm_exporttype_name(export_types.data[i]);
+      module_functions_.insert_or_assign(std::string(name_ptr->data, name_ptr->size), func);
+    } break;
+    case WASM_EXTERN_GLOBAL: {
+      // TODO(mathetake): add support when/if needed.
+    } break;
+    case WASM_EXTERN_MEMORY: {
+      assert(memory_ == nullptr);
+      memory_ = wasm_memory_copy(wasm_extern_as_memory(actual_extern));
+      assert(memory_ != nullptr);
+    } break;
+    case WASM_EXTERN_TABLE: {
+      // TODO(mathetake): add support when/if needed.
+    } break;
+    }
+  }
+
+  wasm_importtype_vec_delete(&import_types);
+  wasm_exporttype_vec_delete(&export_types);
+  wasm_extern_vec_delete(&exports);
+  return true;
+}
+
+static uint32_t parseVarint(const byte_t *&pos, const byte_t *end) {
+  uint32_t n = 0;
+  uint32_t shift = 0;
+  byte_t b;
+  do {
+    if (pos + 1 > end) {
+      abort();
+    }
+    b = *pos++;
+    n += (b & 0x7f) << shift;
+    shift += 7;
+  } while ((b & 0x80) != 0);
+  return n;
+}
+
+std::string_view Wasmtime::getCustomSection(std::string_view name) {
```

Comment:
We should probably extract this code into common, since it's runtime-agnostic. But doing that in subsequent PR is also fine.

Reply:
will do in another PR :+1: